mcp_servers/google_calendar/server.py

```diff
@@ -716,6 +721,238 @@ async def find_free_slots(
+    directory_sources: str = "UNSPECIFIED",
+) -> Dict[str, Any]:
+    """
+    Search for contacts by name or email address.
+
+    Supports searching personal contacts, other contact sources, domain directory,
+    or all sources simultaneously. When contact_type is 'all' (default), returns
+    three separate result sets (personal, other, directory) each with independent
+    pagination tokens.
+    """
+    logger.info(f"Executing tool: search_contacts with query: {query}, contact_type: {contact_type}")
+    try:
+        access_token = get_auth_token()
+        service = get_people_service(access_token)
+
+        # Define the comprehensive read mask for all person fields
+        comprehensive_read_mask = (
+            'addresses,ageRanges,biographies,birthdays,calendarUrls,clientData,coverPhotos,'
+            'emailAddresses,events,externalIds,genders,imClients,interests,locales,locations,'
+            'memberships,metadata,miscKeywords,names,nicknames,occupations,organizations,'
+            'phoneNumbers,photos,relations,sipAddresses,skills,urls,userDefined'
+        )
+
+        # Limited read mask for other contacts
+        limited_read_mask = 'emailAddresses,metadata,names,phoneNumbers'
+
+        def format_contact(person: Dict[str, Any], contact_type_label: str) -> Dict[str, Any]:
+            """Helper function to format a person object into structured contact data."""
+            names = person.get('names', [])
+            emails = person.get('emailAddresses', [])
+            phones = person.get('phoneNumbers', [])
+            orgs = person.get('organizations', [])
+
+            return {
+                'resourceName': person.get('resourceName', ''),
+                'displayName': names[0].get('displayName', 'Unknown') if names else 'Unknown',
+                'firstName': names[0].get('givenName', '') if names else '',
+                'lastName': names[0].get('familyName', '') if names else '',
+                'contactType': contact_type_label,
+                'emailAddresses': [
+                    {
+                        'email': email.get('value', ''),
+                        'type': email.get('type', 'other').lower(),
+                    }
+                    for email in emails
+                ],
+                'phoneNumbers': [
+                    {
+                        'number': phone.get('value', ''),
+                        'type': phone.get('type', 'other').lower(),
+                    }
+                    for phone in phones
+                ],
+                'organizations': [
+                    {
+                        'name': org.get('name', ''),
+                        'title': org.get('title', ''),
+                    }
+                    for org in orgs
+                ],
+                'allFields': person,
+            }
+
+        if contact_type == 'all':
+            # Execute all three searches in parallel
+            import asyncio
```

Comment:
From the official doc, you need to do a warmup cache. See https://developers.google.com/people/api/rest/v1/otherContacts/search.